translate/v34tov33/v34tov33.go

```diff
@@ -0,0 +1,135 @@
+	return
+}
+
+func translateFileEmbedded1(old old_types.FileEmbedded1) (ret types.FileEmbedded1) {
+	tr := translate.NewTranslator()
+	tr.Translate(&old.Append, &ret.Append)
+	tr.Translate(&old.Contents, &ret.Contents)
+	if old.Mode != nil {
+		ret.Mode = util.IntToPtr(*old.Mode)
+	}
+	return
+}
+
+func translateDirectoryEmbedded1(old old_types.DirectoryEmbedded1) (ret types.DirectoryEmbedded1) {
+	if old.Mode != nil {
+		ret.Mode = util.IntToPtr(*old.Mode)
+	}
+	return
+}
+
+func translateLuks(old old_types.Luks) (ret types.Luks) {
+	tr := translate.NewTranslator()
+	tr.AddCustomTranslator(translateTang)
+	tr.Translate(&old.Clevis, &ret.Clevis)
+	tr.Translate(&old.Device, &ret.Device)
+	tr.Translate(&old.KeyFile, &ret.KeyFile)
+	tr.Translate(&old.Label, &ret.Label)
+	tr.Translate(&old.Name, &ret.Name)
+	tr.Translate(&old.Options, &ret.Options)
+	tr.Translate(&old.UUID, &ret.UUID)
+	tr.Translate(&old.WipeVolume, &ret.WipeVolume)
+	return
+}
+
+func translateTang(old old_types.Tang) (ret types.Tang) {
+	tr := translate.NewTranslator()
+	tr.Translate(&old.Thumbprint, &ret.Thumbprint)
+	tr.Translate(&old.URL, &ret.URL)
+	return
+}
+
+func translateConfig(old old_types.Config) (ret types.Config) {
+	tr := translate.NewTranslator()
+	tr.AddCustomTranslator(translateIgnition)
+	tr.AddCustomTranslator(translateDirectoryEmbedded1)
+	tr.AddCustomTranslator(translateFileEmbedded1)
+	tr.AddCustomTranslator(translateLuks)
+	tr.Translate(&old, &ret)
+	return
+}
+
+// end copied Ignition v3_4/translate block
+
+// Translate translates Ignition spec config v3.4 to spec v3.3
+func Translate(cfg old_types.Config) (types.Config, error) {
+	rpt := validate.ValidateWithContext(cfg, nil)
+	if rpt.IsFatal() {
+		return types.Config{}, fmt.Errorf("Invalid input config:\n%s", rpt.String())
+	}
+
+	// 3.3 does not support tang offline provisioning
+	for _, luks := range cfg.Storage.Luks {
+		for _, tang := range luks.Clevis.Tang {
+			if tang.Advertisement != nil && *tang.Advertisement != "" {
+				return types.Config{}, fmt.Errorf("Invalid input config: tang offline provisioning is not supported in spec v3.3")
+			}
+		}
+	}
+	// 3.3 does not support arn: scheme for s3
+	jsonCfg, err := json.Marshal(cfg)
+	if err != nil {
+		return types.Config{}, fmt.Errorf("Failed to marshal config: %v", err)
+	}
+	if strings.Contains(string(jsonCfg), "arn:") {
+		return types.Config{}, fmt.Errorf("Invalid input config: arn: scheme for s3 is not supported in spec v3.3")
+	}
+	// 3.3 does not support luks discard
+	for _, luks := range cfg.Storage.Luks {
+		if luks.Discard != nil && !*luks.Discard {
+			return types.Config{}, fmt.Errorf("Invalid input config: luks discard is not supported in spec v3.3")
+		}
+	}
+	//3.3 does not support luks openOptions
+	for _, luks := range cfg.Storage.Luks {
+		if luks.OpenOptions != nil {
+			return types.Config{}, fmt.Errorf("Invalid input config: luks openOptions is not supported in spec v3.3")
+		}
+	}
+
```

Comment:
Also needs to check for special mode bits.

Reply:
Ok, I was making the assumption that we should not tho. Simply because in igntion we only warn the user if special mode bits are set 3.4 <. So I did not see the value in rejecting a downgrade because it would cause a warning?

Reply:
The warning isn't the issue.  3.4 will apply special mode bits if they are set, and &le; 3.3 will ignore them.

Reply:
Ah I see.